crates/rpcsrv/src/flight/proxy.rs

```diff
@@ -0,0 +1,219 @@
+use crate::errors::{Result, RpcsrvError};
+use crate::proxy::{ProxiedRequestStream, ProxyHandler};
+use crate::util::ConnKey;
+use arrow_flight::flight_service_client::FlightServiceClient;
+use arrow_flight::flight_service_server::FlightService;
+use arrow_flight::{
+    Action, ActionType, Criteria, Empty, FlightData, FlightDescriptor, FlightInfo,
+    HandshakeRequest, HandshakeResponse, PutResult, SchemaResult, Ticket,
+};
+use base64::prelude::*;
+
+use futures::stream::BoxStream;
+use futures::StreamExt;
+use proxyutil::cloudauth::{AuthParams, CloudAuthenticator, ProxyAuthenticator, ServiceProtocol};
+use proxyutil::metadata_constants::{DB_NAME_KEY, ORG_KEY};
+use std::borrow::Cow;
+use std::time::Duration;
+use tonic::{
+    metadata::MetadataMap,
+    transport::{Channel, Endpoint},
+    Status,
+};
+use tonic::{Request, Response, Streaming};
+
+use super::handler::DATABASE_HEADER;
+
+pub type CloudFlightProxyHandler = ProxyHandler<CloudAuthenticator, FlightServiceClient<Channel>>;
+
+impl CloudFlightProxyHandler {
+    async fn connect(&self, meta: &mut MetadataMap) -> Result<FlightServiceClient<Channel>> {
+        let params = Self::auth_params_from_metadata(meta)?;
+        let details = self.authenticator.authenticate(params).await?;
+
+        meta.insert(DATABASE_HEADER, details.database_id.try_into()?);
+        let key = ConnKey {
+            ip: details.ip.clone(),
+            port: details.port.clone(),
+        };
+        // Already have a grpc connection,
+        if let Some(conn) = self.conns.get(&key) {
+            let conn = conn.clone();
+            return Ok(conn);
+        }
+
+        let url = format!("http://{}:{}", key.ip, key.port);
+        let channel = Endpoint::new(url)?
+            .tcp_keepalive(Some(Duration::from_secs(600)))
+            .tcp_nodelay(true)
+            .keep_alive_while_idle(true)
+            .connect()
+            .await?;
+
+        let client = FlightServiceClient::new(channel);
+
+        self.conns.insert(key, client.clone());
+
+        Ok(client)
+    }
+
+    fn auth_params_from_metadata(meta: &MetadataMap) -> Result<AuthParams> {
+        let basic = "Basic ";
+        let authorization = meta
+            .get("authorization")
+            .ok_or_else(|| Status::invalid_argument("authorization field not present"))?
+            .to_str()
+            .map_err(|_| Status::internal("authorization not parsable"))?;
+        if !authorization.starts_with(basic) {
+            Err(Status::invalid_argument(format!(
+                "Auth type not implemented: {authorization}"
+            )))?;
+        }
+        let base64 = &authorization[basic.len()..];
+        let bytes = BASE64_STANDARD
+            .decode(base64)
+            .map_err(|_| Status::internal("authorization not decodable"))?;
+        let s =
+            String::from_utf8(bytes).map_err(|_| Status::internal("authorization not parsable"))?;
+        let parts: Vec<_> = s.split(':').collect();
```

Comment:
the proxy authenticates the session and then passes the authorization information (as key/value pairs, which the server's API should accept. I don't think there's much utility in implementing basic authentication as such, but maybe I'm missing something. 

The (a?) goal state, I think, for this, is:

- the proxy will authenticate with cloud, which will verify a connection with some sort of out-of-bad authentication,
- cloud will generate two certificates:
  - a user certificate, good for a short period of time (circa an hour) which will contain the user/database ID, and will, allow the proxy to (re) authenticate users without communicating with cloud as often. when the cert expires the user will need to get a new cert via either the original password bootstrapping, or we can add a certificate renewal mechanism.
  - a IPC cert, good for much longer (months), which will contain only the DB name and allow the proxy to authenticate to the database (glaredb). 
- glaredb will validate that the IPC cert has been signed by the certificate authority, but will not enforce authorization beyond that. to do this glaredb will receive a copy of CA's cert so that it can validate certificates

Presently, the proxy controls all access to the database workers, and communicates with cloud to ensure that all sessions are validated and correct. Behind the proxy, all connections are trusted, and the proxy has the ability to ensure that authorization information (headers) are correct before passing them to the database workers. Cloud authenticates sessions and hands clients a valid but non-identifying certificate that is used for encryption _only_.

As a fallback, and before before IPC certs are implemented; when the database server does not have a copy of the CA, (as in today's operation) it should fall back to looking for the authorization information in the metadata map. these headers are (I think?) possible to add in at the proxy level (using information from the session authentication (and eventually the user's cert.)

Reply:
I read this a few times, and I don't really understand if this is a question, request for change, or a comment?

Reply:
I wanted to provide some context as to the current state and future of authentication at the glaredb proxy and the glaredb workers. 

I do not think that we should provide basic-style auth headers and instead have the proxy add/set the dbname/etc. in the headers that it sends back to the workers.

Reply:
> why are we implementing http basic auth on this endpoint? 

it's really just from an ergonomic/ usability standpoint. The connectors usually expect an `AUTHORIZATION` header, and it feels weird bypassing this to just stick them into a different header when it's the same information.

```py
conn = connect("grpc+tls://0.0.0.0:6789", db_kwargs={
    DatabaseOptions.AUTHORIZATION_HEADER.value: "Basic <b64_encoded>",
    f"{DatabaseOptions.RPC_CALL_HEADER_PREFIX.value}db_name": "my-db",
    f"{DatabaseOptions.RPC_CALL_HEADER_PREFIX.value}org": "public",
})
```
vs
```py
conn = connect("grpc+tls://0.0.0.0:6789", db_kwargs={
    f"{DatabaseOptions.RPC_CALL_HEADER_PREFIX.value}db_name": "my-db",
    f"{DatabaseOptions.RPC_CALL_HEADER_PREFIX.value}org": "public",
    f"{DatabaseOptions.RPC_CALL_HEADER_PREFIX.value}user": "admin",
    f"{DatabaseOptions.RPC_CALL_HEADER_PREFIX.value}password": "admin",
})
```

> from my understanding, flightsql should (operationally) have the same security model as the RPC service. is this not the case?

yes the security model is exactly the same, we just pass in the user/pass through the `Authorization` header instead of custom headers.

Reply:
- the connectors are going to have to deal with adding the extra headers anyway?
- the `db_name` is the UUID of the deployment (correct?)
- does glaredb look at anything (when in permissive mode?) other than the db name/org name?
- the proxy sets the database name/id?

Reply:
> the connectors are going to have to deal with adding the extra headers anyway?

yes but it's better to follow convention and use the already provided auth header instead of custom headers. We have to use custom headers for `db_name` and `org`, but there are already provided header for user/pass. It feels wrong to use our own headers for a public facing api. RPC is internal so it doesn't really matter, but IMO, the public facing one should try to follow convention as much as possible. We shouldn't follow RPC just because that's the way it was done, we should instead follow convention, and conventional api's will use basic auth header for passing user/pass.  

 
> the db_name is the UUID of the deployment (correct?)

no it's the vanity name 

> does glaredb look at anything (when in permissive mode?) other than the db name/org name?

don't know what you mean by permissive mode. 

> the proxy sets the database name/id?

It set's it on the request headers if that's what you mean.